### ISSUE TITLE
hashmode 22000: nonce-error-corrections is broken

### ISSUE THREAD

**Issue by ZerBea** [hashcat/hashcat#2987]
nonce-error-corrections is broken on BE and LE ANONCEs.
To reproduce run this test hash file, taken from here:
https://hashcat.net/forum/thread-10253-post-53591.html#pid53591

hash line 1:
...1dc265ac ANONCE BE, NC +2 required
hash line 2:
...1fc265ac correct ANONCE
hash line 3:
...1fc265ad ANONCE BE, NC -1 required

```
WPA*02*0e678a6d2a56a24d42592dfd83ae61c2*786256a37308*c8ddc9f59185*e68891e79a84e5a4a7e88fa0e8909d*da74e2ab310bf196035a7567f9d66838e404fc1b1d1c4d62172a4c731dc265ac*0103007502010a0000000000000000000186aca9e4ec2436d207901020d645ab934bca71a399adfb6f9b631e8bf4c19aec000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001630140100000fac040100000fac040100000fac020000*80
WPA*02*0e678a6d2a56a24d42592dfd83ae61c2*786256a37308*c8ddc9f59185*e68891e79a84e5a4a7e88fa0e8909d*da74e2ab310bf196035a7567f9d66838e404fc1b1d1c4d62172a4c731fc265ac*0103007502010a0000000000000000000186aca9e4ec2436d207901020d645ab934bca71a399adfb6f9b631e8bf4c19aec000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001630140100000fac040100000fac040100000fac020000*80
WPA*02*0e678a6d2a56a24d42592dfd83ae61c2*786256a37308*c8ddc9f59185*e68891e79a84e5a4a7e88fa0e8909d*da74e2ab310bf196035a7567f9d66838e404fc1b1d1c4d62172a4c731fc265ad*0103007502010a0000000000000000000186aca9e4ec2436d207901020d645ab934bca71a399adfb6f9b631e8bf4c19aec000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001630140100000fac040100000fac040100000fac020000*80
```

Expected: hashcat recover PSK from all hashes

Unexpected: hashcat disregard NC

```
$ hashcat -m 22000 --nonce-error-corrections=128 test.22000 -a 3 ht3466227
hashcat (v6.2.4-67-gdbefc7e60) starting
0e678a6d2a56a24d42592dfd83ae61c2:786256a37308:c8ddc9f59185:我的大菠萝:ht3466227
                                                          
Session..........: hashcat
Status...........: Exhausted
Hash.Mode........: 22000 (WPA-PBKDF2-PMKID+EAPOL)
Hash.Target......: test.22000
Time.Started.....: Sun Sep 19 00:36:21 2021 (0 secs)
Time.Estimated...: Sun Sep 19 00:36:21 2021 (0 secs)
Kernel.Feature...: Pure Kernel
Guess.Mask.......: ht3466227 [9]
Guess.Queue......: 1/1 (100.00%)
Speed.#1.........:       22 H/s (1.42ms) @ Accel:8 Loops:256 Thr:512 Vec:1
Recovered........: 1/3 (33.33%) Digests
Progress.........: 1/1 (100.00%)
Rejected.........: 0/1 (0.00%)
Restore.Point....: 1/1 (100.00%)
Restore.Sub.#1...: Salt:0 Amplifier:0-1 Iteration:2-5
Candidate.Engine.: Device Generator
Candidates.#1....: ht3466227 -> ht3466227
Hardware.Mon.#1..: Temp: 54c Fan: 32% Util: 92% Core:1885MHz Mem:5005MHz Bus:16
```

**Comment by philsmd** [comment]
just a comment as a hint for @jsteube to troubleshoot this, the problem seems to be here:
https://github.com/hashcat/hashcat/blob/bd2cde31ae659492b41fa8de2bd750b78121a33b/src/modules/module_22000.c#L1074

we would probably need something like this code:
```
          // replaycount not checked, nc needed

          wpa->nonce_error_corrections_chgd = user_options->nonce_error_corrections_chgd;
          wpa->nonce_error_corrections      = user_options->nonce_error_corrections;
```

similar code is in the function `module_hash_binary_parse ()`, but this is only used for binary files.

The problem is that in the `module_hash_decode ()` function we do not have access to `user_options_t *user_options`, because we normally do not need that information in the hash parse function (for other modules).

Setting it to
```
          // replaycount not checked, nc needed

          wpa->nonce_error_corrections_chgd = true;
          wpa->nonce_error_corrections      = 128;
```

worked with my tests, but of course this value must be changeable by the user.

This is just a comment that should help to find a solution faster. thanks

---


**update:**

it's interesting that older versions (without the non-binary support), had similar code but with a hard-coded value, see:
https://github.com/hashcat/hashcat/blob/784eeb257b97b74c6a1143332bb00dc64b7c7474/src/modules/module_22000.c#L740-L742

```
include/types.h:  NONCE_ERROR_CORRECTIONS  = 8
```

**Comment by ZerBea** [comment]
@philsmd good investigation.

Usually hashcat  take this NC information from the message pair field (on demand use default NC +-8 on BE and LE routers, else disable NC [e.g. hcxdumptool will request lost packet and we do not need NC - that will speed up hashcat]).
On most cases default NC +-8 is fine, but I have seen cap files, recorded by a passive dumper, which required a higher NC value, due to heavy packet loss. 
For that, rare cases, it should be overwritable by user option --nonce-error-corrections, similar to hash mode 2500.
hcxpcapngtool will show a notification if hashcat should use a higher NC (e.g. by: --nonce-error-corrections=47) than default:
`REPLAYCOUNT gap (recommended NC).........: 47
`
That worked fine on hash mode 2500.

**Comment by philsmd** [comment]
I think we need to wait for @jsteube to help us with a clever solution (code changes)... but maybe in the meantime you can check if there was ever a version of hashcat (release or git version) that worked with both non-binary hash format and the `--nonce-error-corrections` user option setting. As far as I can see, it never worked... so we need to come up with a new/fixed version/solution

**Comment by ZerBea** [comment]
Yes, it never worked on hashmode 22000, because we tested it only with hashfiles recorded from hcxdumptool (intelligent active dumper that detect packet loss and request lost packet again). That worked as expected and hashcat never ever need to activate NC.
Unfortunately all other tools are passive dumpers (Wireshark, tshark, tcpdump, kismet and airodump-ng which require aireplay-ng or mdk3/4 for the active part) and we have to wrestle against their packet loss, when converting this files to hc22000 hash files.
The cap file, uploaded to the forum is an "absolutely negative" example for this behavior.

**Comment by philsmd** [comment]
maybe we should at least add these 2 lines in the meantime (until we have a solution where the user can also specify the nonce correction value):
```
          wpa->nonce_error_corrections_chgd = true;
          wpa->nonce_error_corrections      = NONCE_ERROR_CORRECTIONS;
```

it seems this is the behaviour that was used [back then](https://github.com/hashcat/hashcat/blob/784eeb257b97b74c6a1143332bb00dc64b7c7474/src/modules/module_22000.c#L740-L742) successfully (the only drawback is that the user can't choose it here, but that also didn't seem to be implemented, nor work, at that time).

**Comment by jsteube** [comment]
Hey there, thanks for the analysis so far. I agree the issue with the nonce-corrections is caused by the plaintext (non-binary) hashlines. But that's only 50% of the problem and we should leave this for architectural change later. This was never working before, only partially because the default was set to 8 instead of 0, but the user specified value was never working.

The more concerning from my perspective is the fact that if you hardcode the 128 here, it only cracks 2/3 hashes. I have debugged the code, but I think hashcat is actually working as expected!

Please check this section:

```
    // now some optimization related to replay counter endianess
    // hcxtools has techniques to detect them
    // since we can not guarantee to get our handshakes from hcxtools we enable both by default
    // this means that we check both even if both are not set!
    // however if one of them is set, we can assume that the endianess has been checked and the other one is not needed

    wpa->detected_le = 1;
    wpa->detected_be = 1;

    if (wpa->message_pair & (1 << 5))
    {
      wpa->detected_le = 1;
      wpa->detected_be = 0;
    }
    else if (wpa->message_pair & (1 << 6))
    {
      wpa->detected_le = 0;
      wpa->detected_be = 1;
    }
```

In the 3 example hashes, the messagepair is set to 0xc0. So we have the 7th and the 6th bit set. The 7th bit enables us to use the replaycounter in general, so this one is fine. However, the 6th bit is set, too! This forces hashcat to only do BE based checking. So hashcat is doing it perfectly as it should. You can verify by hardcoding it to 1, both.

@ZerBea Should we disable this kind of code?

**Comment by ZerBea** [comment]
It is not necessary to to disable that kind of code.
I created the example hashes, mentioned above, by hand and forgot to change the MESSAGEPAIR field according to the BE/LE type of the ANONCE. That is an EPIC fail that caused the other 50% you mentioned.

I fixed my mistake and modified MP field of https://github.com/hashcat/hashcat/issues/2987 to *80
Now we are requesting NC on BE and LE from hashcat:
Recovered........: 1/3 (33.33%) Digests
